From: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Date: Wed, 2 Aug 2023 02:00:23 -0700
Subject: IB/core: Add more speed parsing in ib_get_width_and_speed()
Patch-mainline: v6.6-rc1
Git-commit: ca60fd116c7ee1a4471a8ad0fe07cdfa57f24c11
References: jsc#PED-6864

When the Ethernet driver does not provide the number of lanes
in the __ethtool_get_link_ksettings() response, the function
ib_get_width_and_speed() does not take consideration of 50G,
100G and 200G speeds while calculating the IB width and speed.
Update the width and speed for the above netdev speeds.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Signed-off-by: Selvin Xavier <selvin.xavier@broadcom.com>
Link: https://lore.kernel.org/r/1690966823-8159-1-git-send-email-selvin.xavier@broadcom.com
Signed-off-by: Leon Romanovsky <leon@kernel.org>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 drivers/infiniband/core/verbs.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

--- a/drivers/infiniband/core/verbs.c
+++ b/drivers/infiniband/core/verbs.c
@@ -1899,9 +1899,18 @@ static void ib_get_width_and_speed(u32 n
 		} else if (netdev_speed <= SPEED_40000) {
 			*width = IB_WIDTH_4X;
 			*speed = IB_SPEED_FDR10;
-		} else {
+		} else if (netdev_speed <= SPEED_50000) {
+			*width = IB_WIDTH_2X;
+			*speed = IB_SPEED_EDR;
+		} else if (netdev_speed <= SPEED_100000) {
 			*width = IB_WIDTH_4X;
 			*speed = IB_SPEED_EDR;
+		} else if (netdev_speed <= SPEED_200000) {
+			*width = IB_WIDTH_4X;
+			*speed = IB_SPEED_HDR;
+		} else {
+			*width = IB_WIDTH_4X;
+			*speed = IB_SPEED_NDR;
 		}
 
 		return;
